jupyter_client/tests/test_kernelmanager.py

```diff
@@ -461,3 +461,102 @@ async def test_start_new_async_kernel(self):
             await km.shutdown_kernel(now=True)
             kc.stop_channels()
             self.assertTrue(km.context.closed)
+
+
+class AsyncKernelManagerSubclass(AsyncKernelManager):
+    """Used to test deprecation "routes" that are determined by superclass' detection of methods.
+
+       This class represents a current subclass that overrides both cleanup() and cleanup_resources()
+       in order to be compatible with older jupyter_clients.  We should find that cleanup_resources()
+       is called on these instances vix TestAsyncKernelManagerSubclass.
+    """
+
+    def cleanup(self, connection_file=True):
+        super(AsyncKernelManagerSubclass, self).cleanup(connection_file=connection_file)
+        self.which_cleanup = 'cleanup'
+
+    def cleanup_resources(self, restart=False):
+        super(AsyncKernelManagerSubclass, self).cleanup_resources(restart=restart)
+        self.which_cleanup = 'cleanup_resources'
+
+
+class AsyncKernelManagerWithCleanup(AsyncKernelManager):
+    """Used to test deprecation "routes" that are determined by superclass' detection of methods.
+
+       This class represents the older subclass that overrides cleanup().  We should find that
+       cleanup() is called on these instances via TestAsyncKernelManagerWithCleanup.
+    """
+
+    def cleanup(self, connection_file=True):
+        super(AsyncKernelManagerWithCleanup, self).cleanup(connection_file=connection_file)
+        self.which_cleanup = 'cleanup'
+
+
+class TestAsyncKernelManagerSubclass(AsyncTestCase):
+    def setUp(self):
+        super(TestAsyncKernelManagerSubclass, self).setUp()
+        self.env_patch = test_env()
+        self.env_patch.start()
+
+    def tearDown(self):
+        super(TestAsyncKernelManagerSubclass, self).tearDown()
+        self.env_patch.stop()
+
+    def _install_test_kernel(self):
+        kernel_dir = pjoin(paths.jupyter_data_dir(), 'kernels', 'signaltest')
```

Comment:
Hmm, we may not want to write into their real `jupyter_data_dir` during a test. I would have maybe opted to just mock function call and check that the correct mock is called when a subclass with the old vs new pattern exists. If we do the full integration test pattern like this we likely want to use a temporary directory to cleanup the files automatically and not pollute the system. This might be a bit of a pain to do though :/

Reply:
This install method is used by several existing tests as well.  It doesn't create the kernelspec in the real `jupyter_data_dir` because the [environment setup method](https://github.com/jupyter/jupyter_client/blob/master/jupyter_client/tests/utils.py#L29) points the "important" envs to a temporary directory, knowing that those envs take precedence.

Reply:
Alright that works, thanks for clarifying